bindings/rascal/neighbourlist/structure_manager.py

```diff
@@ -49,18 +52,38 @@ def __init__(self, frames, nl_options, start=None, length=None):
             except Exception as e:
                 raise RuntimeError("Neighbourlist of structures failed "
                                    + "because: " + str(e))
-
-        self.managers = managers
+            self.managers = managers
 
     def __iter__(self):
         return self.managers.__iter__()
 
+    def __len__(self):
+        return len(self.managers)
+
     def __getitem__(self, key):
         return self.managers[key]
 
     def __len__(self):
         return len(self.managers)
 
+    def get_subset(self, selected_ids):
```

Comment:
How so ? The way I see it, the python level parallelization of the kernel computation would be done over the atomic structures (before computing NL) so the output is a numpy array. If we want to do the same but at the representation level, I would run in parallel: build NL, compute representation ; and output the structure manager that would then be accumulated in an AtomsLists. Alternatively you can build the atomsLists and parallelize getting subsets as `get_subset` copies the shared_ptr but the underlying data is the same.
So I am not sure if allowing for too many ways of doing the same thing simplifies the users life. How about directly making small routines that take care of the parallelization ?

Reply:
Ok, that sounds like there are already many ways to do the same thing. Having small routines that wrap the "preferred" way is a good idea, especially since I'm not sure which way is the preferred way right now. The first method (compute in parallel and get numpy arrays) seems especially sub-optimal, especially if you have structures with different sets of atomic species (e.g. QM7b/9) -- how would one join the resulting arrays? There is no optimized dot-product function to compute the kernels in that case.

Do you have a working example of python parallelism over structures with heterogeneous atomic species sets?

Reply:
I do but should I make it part of this PR ?

Reply:
Not necessarily, but I'd like to see it. What branch is it on?

Reply:
there are no branch for it but it can look like that if building the KNM matrix with gradients:

```python
from concurrent.futures import as_completed, ProcessPoolExecutor

def compute(i_frame,frame):
    feat = soap.transform([frame])
    en_row = kernel(feat, X_pseudo)
    grad_rows = kernel(feat, X_pseudo, grad=(True, False))
    return en_row, grad_rows

KNM = np.zeros((Nstructures+Ngrads, X_pseudo.size()))
pbar = tqdm(frames)
with ProcessPoolExecutor(max_workers=20) as executor:
    future_to_compute = {executor.submit(compute, i_frame, frame):i_frame
                                                 for i_frame,frame in enumerate(frames)}
    for future in as_completed(future_to_compute):
        i_frame = future_to_compute[future]
        en_row, grad_rows = future.result()
        KNM[i_frame] = en_row
        KNM[Ngrad_stride[i_frame]:Ngrad_stride[i_frame+1]] = grad_rows
        pbar.update()
```

Reply:
Oh, right. But there's no way to build the pre-sparsified feature matrix in parallel? (i.e. it looks like you have to already have the sparse points to take advantage of this kind of parallelism)

Reply:
Here I compute the KNM matrix so that's what this code snippet does but many things are possible. It also shows the practical advantages of using sparse GPR.

Reply:
Right, but it seems like the current sparsification utilities (the wrapper classes, at least) require a rascal transformed AtomsList, and there's currently no way to get that from a collection of separately computed atoms, is there? I think it would be quite useful to calculate the initial (pre-sparsified) feature matrix in parallel, since this tends to take quite a significant fraction of the time to do a fit. (This is in addition to the issue of doing the sparsification itself in parallel, which may be possible with some conditions.)

Reply:
I am not sure to follow you, are you talking about computing/appending the sparse points in parallel ?

Reply:
No, I'm talking about computing the matrix that you sparsify. That can be quite a large matrix, so it's potentially worth doing in parallel,

Reply:
Isn't that what the code snippet above does ? (Computing the KNM matrix)

Reply:
No, it isn't. I'm talking about the feature matrix that you need to compute because it is the input to the sparsification routines. In order to compute KNM, you need to know what are the M (the sparse points), and in order to know what those are, you need to compute a feature matrix. It is this feature matrix that I would like to be able to compute in parallel.

Reply:
I understand now. It seems quite trivial to change the above snippet to do just that no ? meaning the python parallelization utilities should handle a few use cases.